beacon-chain/blockchain/core.go

```diff
@@ -183,59 +183,148 @@ func (b *BeaconChain) IsEpochTransition(slotNumber uint64) bool {
 	return slotNumber >= b.CrystallizedState().LastStateRecalc()+params.CycleLength
 }
 
-// CanProcessBlock decides if an incoming p2p block can be processed into the chain's block trie.
-func (b *BeaconChain) CanProcessBlock(fetcher types.POWBlockFetcher, block *types.Block) (bool, error) {
+// CanProcessBlockValidator is called by a validator to decide
+// if an incoming p2p block can be processed into the chain's block trie,
+// it checks time stamp, beacon chain parent block hash, and pow chain reference hash.
+func (b *BeaconChain) CanProcessBlockValidator(fetcher types.POWBlockFetcher, block *types.Block) (bool, error) {
 	if _, err := fetcher.BlockByHash(context.Background(), block.PowChainRef()); err != nil {
 		return false, fmt.Errorf("fetching PoW block corresponding to mainchain reference failed: %v", err)
 	}
 
-	// Check if the parentHash pointed by the beacon block is in the beaconDB.
-	parentHash := block.ParentHash()
-	hasParent, err := b.db.Has(parentHash[:])
+	canProcess, err := b.verifyBlockParentHash(block)
 	if err != nil {
-		return false, err
+		return false, fmt.Errorf("unable to process block: %v", err)
 	}
-	// If the block does not have a parent in the database and if that parent is not the genesis block,
-	// then it fails the validity conditions.
-	if !hasParent && block.SlotNumber() != 1 {
-		return false, errors.New("parent hash points to nil in beaconDB")
+	if !canProcess {
+		log.Debug("parent block verification for beacon block %v failed", block.SlotNumber())
+		return false, nil
+	}
+
+	canProcess, err = b.verifyBlockTimeStamp(block)
+	if err != nil {
+		return false, fmt.Errorf("unable to process block: %v", err)
+	}
+	if !canProcess {
+		log.Debug("time stamp verification for beacon block %v failed", block.SlotNumber())
+		return false, nil
+	}
+
+	canProcess, err = b.verifyBlockActiveHash(block)
+	if err != nil {
+		return false, fmt.Errorf("unable to process block: %v", err)
+	}
+	if !canProcess {
+		log.Debug("active state verification for beacon block %v failed", block.SlotNumber())
+		return false, nil
+	}
+
+	canProcess, err = b.verifyBlockCrystallizedHash(block)
+	if err != nil {
+		return false, fmt.Errorf("unable to process block: %v", err)
+	}
+	if !canProcess {
+		log.Debug("crystallized verification for beacon block %v failed", block.SlotNumber())
+		return false, nil
+	}
+	return canProcess, nil
+}
+
+// CanProcessBlockObserver is called by an observer to decide
+// if an incoming p2p block can be processed into the chain's block trie,
+// it checks time stamp, beacon chain parent block hash, and pow chain reference hash.
+func (b *BeaconChain) CanProcessBlockObserver(block *types.Block) (bool, error) {
```

Comment:
So the processing for observer and validator is very similar -

Validator : 
-Check timestamp
-Check active state hashes
-Check crystallized state hashes
-Check beacon parent block's hashes
-Check PoW block reference

Observer validates block by checking the following:

-Check timestamp
-Check active state hashes
-Check crystallized state hashes
-Check beacon parent block's hashes

The only difference is the checking of the POW block hash. can you abstract the common functionality between the validator and observer into a separate function , and then use that in the block processing for both actors. This will reduce duplicated code